src/parquet/bloom_filter.h

```diff
@@ -0,0 +1,165 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_BLOOM_FILTER_H
+#define PARQUET_BLOOM_FILTER_H
+
+#include <cstdint>
+
+#include "parquet/bloom_filter_algorithm.h"
+#include "parquet/hasher.h"
+#include "parquet/types.h"
+#include "parquet/util/memory.h"
+
+namespace parquet {
+class OutputStream;
+
+// A Bloom filter is a compact structure to indicate whether an item is not in a set or
+// probably in a set. Bloom filter class is underlying class of Bloom filter which stores
+// a bit set that represents a set of elements, a hash strategy and a Bloom filter
+// algorithm.
+
+// This Bloom filter uses Murmur3 hash strategy and block-based Bloom filter algorithm.
+class BloomFilter final {
+ public:
+  /// Constructor of Bloom filter. The range of num_bytes should be within
+  /// [MINIMUM_BLOOM_FILTER_BYTES, MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// rounded up/down to lower/upper bound if num_bytes is out of range and also
+  /// will be rounded up to a power of 2. It will use murmur3_x64_128 as its default
+  /// hash function and the block-based algorithm.
+  ///
+  /// @param num_bytes The number of bytes to store Bloom filter bitset.
+  explicit BloomFilter(uint32_t num_bytes);
+
+  /// Construct the Bloom filter with given bit set, it is used when reconstructing
+  /// a Bloom filter from a parquet file. It uses murmur3_x64_128 as its default hash
+  /// function and block-based algorithm as default algorithm.
+  ///
+  /// @param bitset The given bitset to construct Bloom filter.
+  /// @param num_bytes  The number of bytes of given bitset.
+  BloomFilter(const uint32_t* bitset, uint32_t num_bytes);
+
+  BloomFilter(const BloomFilter& orig) = delete;
```

Comment:
A move constructor might be needed for the static deserialization method I mention above.

Reply:
We can return BloomFilter pointer in Deserialization.

Reply:
While that is technically possible, that's a somewhat outdated C++ idiom.